src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+
+                        return TryProcessQueryBody(queryContinuation.Body, processingState);
+                    }
+                }
+
+                return false;
+            }
+
+            private bool TryProcessWhereClause(WhereClauseSyntax whereClause, ProcessingState processingState)
+            {
+                if (IsValidWhereClauseSymbol(whereClause))
+                {
+                    processingState.AddStatementProcessingMethod(
+                         s => SyntaxFactory.Block(
+                             SyntaxFactory.IfStatement(
+                                 ProcessCondition((whereClause).Condition),
+                                 s)));
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool IsValidWhereClauseSymbol(WhereClauseSyntax whereClause)
+            {
+                if (!TryGetMethodSymbol(whereClause, out var methodSymbol))
+                {
+                    return false;
+                }
+
+                var returnType = methodSymbol.ReturnType;
+                if (returnType.OriginalDefinition?.SpecialType != SpecialType.System_Collections_Generic_IEnumerable_T)
+                {
+                    return false;
+                }
+
+                var typeArguments = returnType.GetTypeArguments();
+                if (typeArguments.Length != 1)
+                {
+                    return false;
+                }
+
+                var typeArgument = typeArguments.Single();
+                switch (methodSymbol.Parameters.Length)
+                {
+                    // IEnumerable<T> Where(Func<T, bool> predicate) 
+                    case 1:
+                        return IsPredicate(methodSymbol.Parameters.First().Type, typeArgument);
+
+                    // IEnumerable<T> Where(this IEnumerable<T>, Func<T, bool> predicate) 
+                    case 2:
+                        return
+                            // Compare ReturnType with the first argument
+                            SignatureComparer.Instance.HaveSameSignature(returnType, methodSymbol.Parameters.First(), caseSensitive: true) &&
+                            IsPredicate(methodSymbol.Parameters.Last().Type, typeArgument);
+
+                    default: return false;
+                }
+            }
+
+            private bool TryGetMethodSymbol(QueryClauseSyntax queryClause, out IMethodSymbol methodSymbol)
+            {
+                var queryClauseInfo = _semanticModel.GetQueryClauseInfo(queryClause, _cancellationToken);
+                methodSymbol = queryClauseInfo.OperationInfo.Symbol as IMethodSymbol;
+                return methodSymbol != null;
+            }
+
+            private static bool IsPredicate(ITypeSymbol typeSymbol, ITypeSymbol typeArgument)
+            {
+                if (typeSymbol.Name != "Func")  // TODO better that name check?
```

Comment:
>if (typeSymbol.Name != "Func") [](start = 16, length = 30)

I think we definitely should not be checking names of delegate types.

Reply:
agreed to skip verification of query clauses for now

---
In reply to: [177195620](https://github.com/dotnet/roslyn/pull/25362#discussion_r177195620) [](ancestors = 177195620)